### PR TITLE
Fixing decode error of + strings (plus sign)

### PR DESCRIPTION
**Pull request by TBeijen**
Fix for #13

The first regex OR (now removed) caused regex to pass even if none of the succeeding oct|dec|hex definitions was present.

Quite confident this fixes it:
* Testsuite contains tests for oct|dec|hex (`decode_test.go`)
* No unexpected side-effects in testsuite

### CODE REVIEW

**Comment by apparentlymart** [comment]
Hi @TBeijen! Thanks for working on this and sorry for the very slow response.

I agree that this seems like a more accurate read of the rules in [Tag Resolution](https://yaml.org/spec/1.2-old/spec.html#id2805071): it requires there to be at least one digit after the `+` or `-` to count as an `tag:yaml.org,2002:int`, but this was previously allowing zero digits.

Looking closely at that specification text, I think there's also another inconsistency: the `[-+]?` part is included only for the decimal form, and not for the octal and hex forms. Therefore I think the `[-+]?` pattern ought to be _inside_ the disjunction directly before the `[0-9_]+` pattern.

In other words, I think the correct pattern is `(0o[0-7_]+|[-+]?[0-9_]+|0x[0-9a-fA-F_]+)`. Would you agree?

**Comment by TBeijen** [comment]
Interesting. I overlooked the divergence from YAML spec, just reasoned "will also match if neither decimal, octal or hexadecimal value follows the sign. That's no good." This is what code reviews are for! 👌 

To some extent I would also be inclined to move towards the YAML spec, so what you proposed by bringing `[-+]?` inside the disjunction.

But:

* This PR then becomes a breaking change as `TestUnmarshal/single_hexadecimal_int_negative_implied_by_parsability` then fails.
* Apparently YAML spec does not provide way to specify negative octal or hexadecimal ints? Am I overlooking something? Various online converters for decimal to octal suggest prefixing with `-` is valid, same as decimal.
* Probably there has been good reason to diverge from the YAML spec, allowing `[-+]?` for base8 and base16 as well (since link to spec has been added in that exact PR). But that context is now lost. What are we missing?

Likewise for the `_`: 

* Removing that fails `TestUnmarshal/single_hexadecimal_int_with_underscores_implied_by_parsability`.
* Although underscore support added for base8, base10 and base16, test coverage is incomplete (just base16)
* What context do we miss about reason it has been added?

So, bottom line, I would favor creating a separate issue along the lines of "Bring int parsing in accordance to YAML spec 10.3.2 (breaking)", and proceed with this non-breaking PR as is. Based on tagging practices, this PR could be a `1.0.4`, breaking compatibility would imply `1.1.0` or even `2.0.0`.

**Comment by apparentlymart** [comment]
Thanks for that additional thought and consideration!

I guess from my perspective the intention of this library is to implement the YAML specifications and so interpreting something differently than the spec requires is a bug to be fixed even if it was previously allowing invalid YAML to be decoded.

Of course there is some subtlety in that argument. I imagine fixing this bug would cause some input that previously produced a number to begin producing a string instead, which indeed might upset existing assumptions. But interpretation of a scalar as a number is inherently lossy -- it discards the exact characters used in the scalar -- so producing a string seems like the better treatment when author intent is unclear, since it's always possible to perform the string-to-number conversation later, but it's impossible to recover the original string from a number.

Therefore I think my conclusion here is that it's better to honor the rules described in the YAML spec rather than treating this library's current bugs as a de-facto YAML extension, even at the expense of risking changing the interpretation of existing input.

I don't know why YAML doesn't support negative numbers in other bases, but supporting it when the YAML spec says not to implies that some valid YAML input would be unreadable using this library (if the author was expecting the string interpretation required by the spec) so I think it's preferable to behave how the spec requires than require users of this library to learn about undocumented differences relative to YAML.

If we're going to change anything here then I'd prefer the changelog to say that interpretation of scalars as numbers now matches the YAML spec completely, rather than just that we've nudged it only very slightly in that direction, since that way we get everything fixed in one step rather than having a confusing middle period where it's still incorrect but just slightly less incorrect.

**Comment by TBeijen** [comment]
> Therefore I think my conclusion here is that it's better to honor the rules described in the YAML spec rather than treating this library's current bugs as a de-facto YAML extension, even at the expense of risking changing the interpretation of existing input.

Agreed that adhering to YAML spec should (ultimately) be the goal. It removes all ambiguity, especially since spec already provides the regular expressions to use.

If something breaks, it will be edge cases that didn't follow YAML spec to begin with.

I'll update the PR somewhere in coming days.

**Comment by TBeijen** [comment]
@apparentlymart It took a bit more than 'the coming days' but finally updated the PR.

Int parsing now exactly follows the YAML spec and uses exact same regex patters as outlined in [YAML spec section 10.3.2](https://yaml.org/spec/1.2-old/spec.html#id2805071)

> I guess from my perspective the intention of this library is to implement the YAML specifications

Nice thoughtful discussion so far, great stuff! And although I initially proposed another approach, I agree that the above essentially _is_ the library goal. The current PR will remove confusion and friction, even though some previously supported edge cases are now dropped.

**Comment by apparentlymart** [comment]
Hi @TBeijen! Sorry I went quiet again. I've been away from this project for a few months but I'm returning now.

The latest version of this looks great to me and I'm going to merge it now. I'm going to hold off on tagging a numbered release for the moment since I anticipate there will be some other things for me to catch up on that could potentially get grouped together into a single release, but I'm not 100% sure yet since I'm still reloading the context.

Thanks for working on this, and for your patience in working through the tradeoffs with me.

**Comment by apparentlymart** [comment]
Hi again @TBeijen,

I just wanted to followup and let you know that today I've tagged v1.1.0 which includes this fix. Thanks again!

**Comment by TBeijen** [comment]
> Hi again @TBeijen,
> 
> I just wanted to followup and let you know that today I've tagged v1.1.0 which includes this fix. Thanks again!

Awesome! Thanks.
